// @kinde-oss/kinde-node-express/dist-cjs/setup/kindeClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupInternalClient = exports.validateInitialConfig = exports.getInternalClient = exports.getInitialConfig = void 0;
const kinde_typescript_sdk_1 = require("@kinde-oss/kinde-typescript-sdk");
const version_1 = require("../version");
let initialConfig;
let internalClient;
const commonConfigParams = [
    'grantType',
    'issuerBaseUrl',
    'postLogoutRedirectUrl',
    'siteUrl',
    'clientId',
];
const grantTypeConfigParams = {
    [kinde_typescript_sdk_1.GrantType.AUTHORIZATION_CODE]: ['secret', 'redirectUrl', 'unAuthorisedUrl'],
    [kinde_typescript_sdk_1.GrantType.PKCE]: ['redirectUrl', 'unAuthorisedUrl'],
    [kinde_typescript_sdk_1.GrantType.CLIENT_CREDENTIALS]: ['secret'],
};
/**
 * Returns the above `initialConfig` private to this module, throws an exception if
 * said `initialConfig` has not been initialized.
 *
 * @returns {SetupConfig}
 */
const getInitialConfig = () => {
    if (initialConfig === undefined) {
        throw new Error('Initial config is not initialized');
    }
    else {
        return initialConfig;
    }
};
exports.getInitialConfig = getInitialConfig;
/**
 * Returns the above `internalClient` private to this module, throws an exception if
 * said `internalClient` has not been initialized.
 *
 * @returns {ClientType<G>}
 */
const getInternalClient = () => {
    if (internalClient === undefined) {
        throw new Error('Internal Kinde server client is not initialized');
    }
    else {
        return internalClient;
    }
};
exports.getInternalClient = getInternalClient;
/**
 * Validates provided config, raises exception if any required parameters are
 * missing otherwise returns same config.
 *
 * @param {SetupConfig<G>} config
 * @returns {SetupConfig<G>}
 */
const validateInitialConfig = (config) => {
    const grantType = config.grantType;
    const grantTypeParams = grantTypeConfigParams[grantType];
    if (!grantTypeParams) {
        const types = Object.values(kinde_typescript_sdk_1.GrantType).join(', ');
        throw new Error(`Provided grant type must be one of these values ${types}`);
    }
    const configParams = [...commonConfigParams, ...grantTypeParams];
    const configObject = config;
    configParams.forEach((param) => {
        const value = configObject[param];
        if (!value || typeof value !== 'string') {
            throw new Error(`Required config parameter '${param}' has invalid value ${value}`);
        }
    });
    return config;
};
exports.validateInitialConfig = validateInitialConfig;
/**
 * Initializes above `initialConfig` and `internalClient` raises an exception if
 * validation of provided config failes, returns created client otherwise.
 *
 * @param {SetupConfig<G>} config
 * @returns {ClientType<G>}
 */
const setupInternalClient = (config) => {
    const { issuerBaseUrl, redirectUrl, postLogoutRedirectUrl, audience, scope, secret, clientId, } = config;
    initialConfig = (0, exports.validateInitialConfig)(config);
    internalClient = (0, kinde_typescript_sdk_1.createKindeServerClient)(initialConfig.grantType, {
        authDomain: issuerBaseUrl,
        redirectURL: redirectUrl,
        clientId: clientId ?? 'reg@live',
        clientSecret: secret,
        logoutRedirectURL: postLogoutRedirectUrl,
        audience: audience ?? undefined,
        scope: scope ?? undefined,
        framework: 'ExpressJS',
        frameworkVersion: version_1.version,
    });
    return internalClient;
};
exports.setupInternalClient = setupInternalClient;
